refresh.template.py

```diff
@@ -212,36 +215,105 @@ def _get_cpp_command_for_files(compile_action: json):
+# Yields the entries to compile_commands.json
+def get_commands(directory: pathlib.Path, target: str, flags: str) -> str:
+    # Log clear completion messages
+    print(f"\033[0;34m>>> Analyzing commands used in {target}\033[0m")
+
+    # Queries Bazel's C-family compile actions, and runs them through our extract.py reformatter
+    # Aquery docs if you need em: https://docs.bazel.build/versions/master/aquery.html
+    # One bummer, not described in the docs, is that aquery filters over *all* actions for a given target, rather than just those that would be run by a build to produce a given output. This mostly isn't a problem, but can sometimes surface extra, unnecessary, misconfigured actions. Chris has emailed the authors to discuss and filed an issue so anyone reading this could track it: https://github.com/bazelbuild/bazel/issues/14156.
+    # We switched to jsonproto instead of proto because of https://github.com/bazelbuild/bazel/issues/13404. We could change back when fixed--reverting most of the commit that added this line and tweaking the build file to depend on the target in that issue. That said, it's kinda nice to be free of the dependency, unless (OPTIMNOTE) jsonproto becomes a performance bottleneck compated to binary protos.
+
+    cmd = [
+        "bazel",
+        "aquery",
+        f"mnemonic('(Objc|Cpp)Compile',deps({target}))",
+        "--ui_event_filters=-info",
+        "--noshow_progress",
+        "--output=jsonproto",
+    ] + shlex.split(flags)
+
+    completed = subprocess.run(
+        cmd,
+        cwd=os.fspath(directory),
+        encoding="utf-8",
+        errors="replace",
+        check=True,
+        stdout=subprocess.PIPE,
+    )
+
+    # Load aquery's output from the proto data being piped to stdin
+    # Proto reference: https://github.com/bazelbuild/bazel/blob/master/src/main/protobuf/analysis_v2.proto
+    yield from extract(
+        directory,
+        # object_hook allows object.member syntax, just like a proto, while avoiding the protobuf dependency
+        json.loads(completed.stdout, object_hook=lambda d: SimpleNamespace(**d)),
+    )
+
+    # Log clear completion messages
+    print(f"\033[0;32m>>> Finished extracting commands for {target}\033[0m")
+
+
+def get_all_commands(directory: pathlib.Path) -> typing.Iterable[str]:
+    args = [
+        # Begin: Command template filled by Bazel
+        # {get_commands}
+        # End: Command template filled by Bazel
+    ]
+    for (target, flags) in args:
+        yield from get_commands(directory, target, flags)
+
+
+if __name__ == "__main__":
+    # Get workspace's root so compile_commands.json goes the right place
+    # -- and so we can invoke Bazel on the repo within this script.
+    directory = pathlib.Path(os.environ["BUILD_WORKSPACE_DIRECTORY"])
+    commands = get_all_commands(directory)
+
+    # Chain output into compile_commands.json
+    with open(directory / "compile_commands.json", "w") as fob:
+        json.dump(list(commands), fob)
```

Comment:
Whoops! Just realized: Can we keep the old indent=2, check_circular=False, for readability and speed, respectively? 

[Bed time over here, but thanks so much for looking through!]

Reply:
Done